File: src/main/java/com/chatchatabc/jpademojava/impl/domain/service/CountryServiceImpl.java
package com.chatchatabc.jpademojava.impl.domain.service;

import com.chatchatabc.jpademojava.domain.model.Country;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.CountryRepository;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.CountryService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class CountryServiceImpl implements CountryService {
    @Autowired
    private CountryRepository countryRepository;
    @Autowired
    private UserRepository userRepository;

    /**
     * Create country
     *
     * @param country
     * @return
     */
    @Override
    public Country create(Country country) {
        return countryRepository.save(country);
    }

    /**
     * Update country
     *
     * @param countryId
     * @param newCountryInfo
     * @return
     */
    @Override
    public Country update(String countryId, Country newCountryInfo) throws Exception {
        Optional<Country> queriedCountry = countryRepository.findById(countryId);
        if (queriedCountry.isEmpty()) {
            throw new Exception("Country not found");
        }

        // Update Fields
        if (newCountryInfo.getName() != null) {
            queriedCountry.get().setName(newCountryInfo.getName());
        }

        return countryRepository.save(queriedCountry.get());
    }

    /**
     * Assign user to country
     *
     * @param userId
     * @param countryId
     * @return
     */
    @Override
    public User assign(String userId, String countryId) throws Exception {
        Optional<User> user = userRepository.findById(userId);
        if (user.isEmpty()) {
            throw new Exception("User not found");
        }
        Optional<Country> country = countryRepository.findById(countryId);
        if (country.isEmpty()) {
            throw new Exception("Country not found");
        }

        // Apply Update
        user.get().setCountry(country.get());

        return userRepository.save(user.get());
    }
}

File: src/main/kotlin/com/chatchatabc/jpademo/impl/domain/service/CountryServiceImpl.kt
package com.chatchatabc.jpademo.impl.domain.service

import com.chatchatabc.jpademo.application.dto.country.CountryUpdateRequest
import com.chatchatabc.jpademo.domain.model.Country
import com.chatchatabc.jpademo.domain.model.User
import com.chatchatabc.jpademo.domain.repository.CountryRepository
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.CountryService
import org.springframework.stereotype.Service

@Service
class CountryServiceImpl (
    private val countryRepository: CountryRepository,
    private val userRepository: UserRepository
) : CountryService {

    /**
     * Create Country
     */
    override fun create(country: Country): Country {
        return countryRepository.save(country)
    }

    /**
     * Update Country
     */
    override fun update(countryId: String, request: CountryUpdateRequest): Country {
        val queriedCountry = countryRepository.findById(countryId)
        if (queriedCountry.isEmpty) {
            throw Exception("Country not found")
        }

        // Update fields
        queriedCountry.get().apply {
            name = request.name ?: name
        }
        return countryRepository.save(queriedCountry.get())
    }

    /**
     * Assign Country to User
     */
    override fun assign(userId: String, countryId: String): User {
        val user = userRepository.findById(userId)
        if (user.isEmpty) {
            throw Exception("User not found")
        }
        val country = countryRepository.findById(countryId)
        if (country.isEmpty) {
            throw Exception("Country not found")
        }

        // Apply Update
        user.get().apply {
            this.country = country.get()
        }

        return userRepository.save(user.get())
    }

    /**
     * Unassign Country from User
     */
    override fun unassign(userId: String): User {
        val user = userRepository.findById(userId)
        if (user.isEmpty) {
            throw Exception("User not found")
        }
        user.get().country = null
        return userRepository.save(user.get())
    }


}